sumcheck_macro/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+extern crate proc_macro;
+use proc_macro2::Span;
+use quote::quote;
+use syn::{
+    Expr, ExprClosure, Ident, LitInt, Result, Token,
+    parse::{Parse, ParseStream},
+    parse_macro_input,
+};
+
+struct MyMacroInput {
+    number: LitInt,
+    product_access: ExprClosure,
+}
+
+impl Parse for MyMacroInput {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let number = input.parse()?;
+        input.parse::<Token![,]>()?;
+        let expr = input.parse()?;
+        match expr {
+            Expr::Closure(product_access) => Ok(Self {
+                number,
+                product_access,
+            }),
+            _ => Err(syn::Error::new_spanned(
+                expr,
+                "Expected closure that gives access to the mle product",
+            )),
+        }
+    }
+}
+
+#[allow(unused_macros)]
+#[proc_macro]
+pub fn sumcheck_code_gen(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let input = parse_macro_input!(input as MyMacroInput);
+
+    let degree = input.number.base10_parse::<u32>().unwrap();
+    let product_access = input.product_access;
+
+    // Output: let product_access = |i| closure_raw_code;
+    let mut out = quote! {
+        let product_access = #product_access;
+    };
+
+    // Declare vars to access flattened_ml_extensions from products
+    // let f0 = product_access(0u32);
+    // let f1 = product_access(1u32);
+    // ...
+    let mut f_var_names = Vec::new();
+    for i in 1..=degree {
+        let f_var_name = ident(format!("f{}", i));
+        let idx = (i - 1) as usize;
+        out = quote! {
+            #out
+            let #f_var_name = product_access(#idx);
+        };
+        f_var_names.push(f_var_name);
+    }
+
+    // Map flattened_ml_extensions evaluations
+    // match (&f0.evaluations(), &f1.evalutations(), ...) {
+    //     (FieldType::Base(base1), FieldType::Base(base2), ...) => {...}
+    //     ...
+    // }
+    let match_input = f_var_names
+        .iter()
+        .fold(proc_macro2::TokenStream::new(), |acc, f| {
+            if acc.is_empty() {
+                quote! {&#f.evaluations()}
+            } else {
+                quote! {#acc, &#f.evaluations()}
+            }
+        });
+
+    // There are 2^degree possible combinations of Base and Ext evaluations, but
+    // since it is commutative, we only need to consider degree+1 combinations, hence
+    // we have to forward the rest of the combinations to the default case. The way it
+    // is done here is by sorting.
+    let mut sorter_match_arms = proc_macro2::TokenStream::new();
+    for case in 0..(2u32.pow(degree)) {
+        // 0 -> Base
+        // 1 -> Ext
+        let bits_og = (0..degree)
+            .enumerate()
+            .map(|(idx, shift)| (idx, (case >> shift) & 1))
+            .collect::<Vec<_>>();
+
+        let mut bits_sorted = bits_og.clone();
+        bits_sorted.sort_by(|a, b| a.1.cmp(&b.1));
+
+        let is_sorted = bits_sorted
```

Comment:
You can use itertools's [`tuple_windows`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.tuple_windows) to make this much cleaner.

Reply:
I've made the change. 

Yes, you are right. In the first match statement, I want to include only those cases that are sorted, I want to ignore cases that were by default sorted. I've also added comments for more context on what the code is trying to generate.